File: algorithms/cpplus/CountSort.cpp
#include "Driver.h"

template <typename T>
void countSort(std::vector<T>& input, T minimal, T maximal)
{
  std::vector<int> count(maximal - minimal + 1);
  std::vector<T> output(input.size());
  for (const auto& i : input) {
    ++count[i];
  }
  for (int i = 1; i < count.size(); ++i) {
    count[i] += count[i - 1];
  }
  for (const auto& i : input) {
    output[count[i] - 1] = i;
    --count[i];
  }
  input = output;
}

void doCountSort(std::vector<int>& input)
{
  countSort(input, 0, 100);
}

int main(int argn, char* argv[]) 
{
  runMethod(doCountSort);
  return 0;
}

File: algorithms/sorting/java/CountSort.java
class CountSort implements SortMethod
{
  private int minimal;
  private int maximal;

  public CountSort(int minimal, int maximal)
  {
    this.minimal = minimal;
    this.maximal = maximal;
  }

  @Override
  public void sort(int[] input)
  {
    int[] count = new int[maximal - minimal + 1];
    int[] output = new int[input.length];
    for (int i : input) {
      ++count[i];
    }
    for (int i = 1; i < count.length; ++i) {
      count[i] += count[i - 1];
    }
    for (int i : input) {
      output[count[i] - 1] = i;
      --count[i];
    }
    System.arraycopy(output, 0, input, 0, input.length);
  }

  public static void main(String[] argv)
  {
    Driver.run(new CountSort(0, 100)); 
  }
}
